src/ucs/datastruct/lru.c

```diff
@@ -0,0 +1,68 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2023. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/datastruct/list.h>
+#include <ucs/datastruct/lru.h>
+#include <ucs/debug/memtrack_int.h>
+
+
+ucs_status_t ucs_lru_create(ucs_lru_h *lru_p, size_t capacity)
+{
+    ucs_lru_h lru;
+
+    if (capacity == 0) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    lru = ucs_calloc(1, sizeof(*lru), "ucs_lru");
+    if (lru == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    kh_init_inplace(ucs_lru_hash, &lru->hash);
+
+    /* Resize the cache to the required capacity. Need to allocate extra space
+     * in order to avoid collisions in the hash table. */
+    if (kh_resize(ucs_lru_hash, &lru->hash, capacity * 2) < 0) {
+        ucs_free(lru);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_list_head_init(&lru->list);
+
+    lru->capacity = capacity;
+    lru->size     = 0;
+    *lru_p        = lru;
+    return UCS_OK;
+}
+
+void ucs_lru_destroy(ucs_lru_h lru)
+{
+    kh_destroy_inplace(ucs_lru_hash, &lru->hash);
+    ucs_free(lru);
+}
+
+ucs_status_t ucs_lru_get_all(const ucs_lru_h lru, ucs_lru_array_t *array)

```

Comment:
maybe use callback function (called for each key), or "foreach" style macro?
so would not assume the user wants to put all items in an array